crates/net/src/eventsource/futures.rs

```diff
@@ -0,0 +1,258 @@
+//! es.subscribe_event("some-event-type").unwrap();
+//! es.subscribe_event("another-event-type").unwrap();
+//!
+//! spawn_local(async move {
+//!     while let Some(Ok((event_type, msg))) = es.next().await {
+//!         console_log!(format!("1. {}: {:?}", event_type, msg))
+//!     }
+//!     console_log!("EventSource Closed");
+//! })
+//! # }
+//! ```
+use crate::eventsource::{EventSourceError, State};
+use crate::js_to_js_error;
+use futures_channel::mpsc;
+use futures_core::{ready, Stream};
+use gloo_utils::errors::JsError;
+use pin_project::{pin_project, pinned_drop};
+use std::collections::hash_map::Entry;
+use std::collections::HashMap;
+use std::ops::DerefMut;
+use std::pin::Pin;
+use std::sync::{Arc, Mutex};
+use std::task::{Context, Poll};
+use wasm_bindgen::prelude::*;
+use wasm_bindgen::JsCast;
+use web_sys::MessageEvent;
+
+/// Wrapper around browser's EventSource API.
+#[allow(missing_debug_implementations)]
+#[pin_project(PinnedDrop)]
+pub struct EventSource {
+    es: web_sys::EventSource,
+    message_sender: mpsc::UnboundedSender<StreamMessage>,
+    #[pin]
+    message_receiver: mpsc::UnboundedReceiver<StreamMessage>,
+    #[allow(clippy::type_complexity)]
+    closures: Arc<
+        Mutex<(
+            HashMap<String, Closure<dyn FnMut(MessageEvent)>>,
+            Closure<dyn FnMut(web_sys::Event)>,
+        )>,
+    >,
+}
+
+impl EventSource {
+    /// Establish an EventSource.
+    ///
+    /// This function may error in the following cases:
+    /// - The connection url is invalid
+    ///
+    /// The error returned is [`JsError`]. See the
+    /// [MDN Documentation](https://developer.mozilla.org/en-US/docs/Web/API/EventSource/EventSource#exceptions_thrown)
+    /// to learn more.
+    pub fn new(url: &str) -> Result<Self, JsError> {
+        let es = web_sys::EventSource::new(url).map_err(js_to_js_error)?;
+
+        let (message_sender, message_receiver) = mpsc::unbounded();
+
+        let error_callback: Closure<dyn FnMut(web_sys::Event)> = {
+            let sender = message_sender.clone();
+            Closure::wrap(Box::new(move |e: web_sys::Event| {
+                let sender = sender.clone();
+                let is_connecting = e
+                    .current_target()
+                    .and_then(|target| target.dyn_into::<web_sys::EventSource>().ok())
+                    .map(|es| es.ready_state() == web_sys::EventSource::CONNECTING)
+                    .unwrap_or(false);
+                if !is_connecting {
+                    let _ = sender.unbounded_send(StreamMessage::ErrorEvent);
+                };
+            }) as Box<dyn FnMut(web_sys::Event)>)
+        };
+
+        es.set_onerror(Some(error_callback.as_ref().unchecked_ref()));
+
+        Ok(Self {
+            es,
+            message_sender,
+            message_receiver,
+            closures: Arc::new(Mutex::new((HashMap::new(), error_callback))),
+        })
+    }
+
```

Comment:
```suggestion
    pub fn subscribe_all(&mut self) -> Result<(), JsError> { ... }
```
This will listen to all events emitted (`onmessage`)

Reply:
I do not believe this can be implemented. I believe that registering the `on_message` handler will return just those events with no event type.

Reply:
@huntc is right, according to the spec registering the `on_message` handler only listens to unnamed events, i.e. these without type.